app/src/main/java/com/github/polypoly/app/models/game/GameViewModel.kt

```diff
@@ -43,10 +63,160 @@ class GameViewModel(
+        if (playerStateData.value != expectedFrom)
+            return
+        playerStateData.value = to
+    }
+
+    /**
+     * Ends ROLLING_DICE state and moves to MOVING
+     */
+    fun diceRolled() {
+        playerStateFSMTransition(PlayerState.ROLLING_DICE, PlayerState.MOVING)
+    }
+
+    /**
+     * Ends MOVING state and moves to INTERACTING
+     */
+    fun locationReached() {
+        playerStateFSMTransition(PlayerState.MOVING, PlayerState.INTERACTING)
+    }
+
+    /**
+     * Ends INTERACTING state and moves to BETTING
+     */
+    fun startBetting() {
+        playerStateFSMTransition(PlayerState.INTERACTING, PlayerState.BETTING)
+    }
+
+    /**
+     * Ends BETTING state and moves back to INTERACTING
+     */
+    fun cancelBetting() {
+        playerStateFSMTransition(PlayerState.BETTING, PlayerState.INTERACTING)
+    }
+
+    /**
+     * Resets player state back to the beginning of a turn (ie ROLLING_DICE)
+     */
+    fun resetTurnState() {
+        playerStateData.value = PlayerState.ROLLING_DICE
+    }
+
+    /**
+     * Computes the closest location to the given position among the existing locations in the game.
+     * @param position Reference position
+     * @return A future holding the closest location found or null if the closest location is farther than MAX_INTERACT_DISTANCE
+     */
+    fun computeClosestLocation(position: GeoPoint): CompletableFuture<LocationProperty?> {
+        val result = CompletableFuture<LocationProperty?>()
+
+        viewModelScope.launch {
+            var closestLocation: LocationProperty? = null
+            var closestDistance = Double.MAX_VALUE
+
+            val allLocations = gameData.value?.allLocations ?: listOf()
+            for (location in allLocations) {
+                val distance = position.distanceToAsDouble(location.position())
+                if (distance < closestDistance) {
+                    closestLocation = location
+                    closestDistance = distance
+                }
+            }
+
+            if (closestDistance > MAX_INTERACT_DISTANCE) {
+                closestLocation = null
+            }
+
+            result.complete(closestLocation)
+        }
+
+        return result
+    }
+
+    /**
+     * Rolls the dice and returns the location that corresponds to the sum of 2 dice rolls, 3 times
+     * ensuring that the player does not visit the same location twice.
+     */
+    fun rollDiceLocations(currentLocation: LocationProperty?): CompletableFuture<List<LocationProperty>> {
+        val result = CompletableFuture<List<LocationProperty>>()
+
+        viewModelScope.launch {
+            val locationsNotToVisitName = mutableListOf<String>()
+            if (currentLocation != null)
+                locationsNotToVisitName.add(currentLocation.name)
+
+            val allLocations = gameData.value?.allLocations ?: listOf()
+
+            val locationsToVisit = mutableListOf<LocationProperty>()
+            for (i in 1..3) {
```

Comment:
Maybe we should put global constants specific to the whole game and that one of these constants would be `numberOfLocationProposePerTurn = 3`